pkg/util/slidingwindow/sliding_window.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package slidingwindow
+
+import "time"
+
+// Swag represents a sliding window aggregator over a binary operation.
+// The aggregator will aggregate recorded values in two ways:
+// (1) Within each window, the binary operation is applied when recording a new
+//     value into the current window.
+// (2) On Query, the binary operation is accumulated over every window, from
+//     most to least recent.
+// The binary operator function must therefore be:
+//     associative :: binOp(binOp(a,b), c) = binOp(a,binOp(b,c))
+// In order to have correct results. Note that this does not allow for a more
+// general class of aggregators that are associative, such as geometric mean,
```

Comment:
They may be associative with the following additional operations available!

They would require actually tracking every value in the sliding window rather than just aggregating in X duration windows. (for lift() lower())